torch/_decomp/decompositions.py

```diff
@@ -1225,6 +1225,83 @@ def cudnn_batch_norm_backward(
     )
 
 
+@register_decomposition(aten._adaptive_avg_pool2d)
+def adaptive_avg_pool2d(input: Tensor, output_size: Tuple[int, int]):
+    # Preconditions
+    device = input.device
+    shape = input.shape
+    ndim = len(shape)
+    utils.check(
+        ndim in (3, 4),
+        lambda: f"adaptive_avg_pool2d(): Expected 3D or 4D tensor, but got {ndim}",
+    )
+    for d in input.shape[-2:]:
+        utils.check(
+            d != 0,
+            lambda: "adaptive_avg_pool2d(): Expected input to have non-zero size for "
+            f"non-batch dimensions, but input has shape {tuple(shape)}.",
+        )
+
+    # Optimisation (we should also do this in the kernel implementation)
+    if shape[-2] % output_size[-2] == 0 and shape[-1] % output_size[-1] == 0:
+        stride = tuple(i // o for i, o in zip(shape[-2:], output_size))
+        kernel = tuple(
+            i - (o - 1) * s for i, o, s in zip(shape[-2:], output_size, stride)
+        )
+        return torch.nn.functional.avg_pool2d(input, kernel, stride)
+
+    def start_index(a, b, c):
+        return (a * c).float().div(b).floor().long()
```

Comment:
converting to float could introduce rounding issues.

Reply:
That's a good point in abstract, but I doubt it is an issue in practice, as for that to happen you'd need to have a dimension with size 16M, but fair enough.
This is translated from the implementation in core, so we should probably fix it there as well.